# Tidyverse data wrangling

```{r, include=FALSE}

collapsible_code <- function(html_content, non_html_content) {
  output_format <- knitr::opts_knit$get("rmarkdown.pandoc.to")
  # opts_knit$get("out.format") is not reliable for determining the output format
  if (output_format == "html") {
    # HTML output (GitBook/HTML): Show collapsible section
    cat(html_content)
  } else {
    # Non-HTML output (EPUB, PDF): Show regular R code block
    cat(non_html_content)
  }
}

```


[Tidyverse](https://www.tidyverse.org/packages/) is a suite of packages that are incredibly useful for working with data. are designed to work together to make common data science operations more user friendly. The packages have functions for data wrangling, tidying, reading/writing, parsing, and visualizing, among others. 

![](img/tidyverse_website.png)

## What is data wrangling?
Data wrangling: The process of cleaning and unifying messy and complex data sets for easy access and analysis.

![](img/wrangling.png)

## Tidyverse basics

**Two important new concepts in tidyverse that we will focus on are pipes and tibbles**.

Before we get started with pipes or tibbles, let's load the library:

```{r}
library(tidyverse)
```

### Pipes

To make R code more human readable, the Tidyverse tools use the pipe, `%>%`, which is part of the `dplyr` package that is installed automatically with Tidyverse. **The pipe allows the output of a previous command to be used as input to another command instead of using nested functions.**

>**NOTE:** Shortcut to write the pipe is <kbd>shift</kbd> + <kbd>command</kbd> + <kbd>M</kbd> in MacOS; for Windows <kbd>shift</kbd> + <kbd>ctrl</kbd> + <kbd>M</kbd>

An example of using the pipe to run multiple commands:

```{r}

## A single command
sqrt(83)

## Base R method of running more than one command
round(sqrt(83), digits = 2)

## Running more than one command with piping
sqrt(83) %>% round(digits = 2)
```

The pipe represents a much easier way of writing and deciphering R code.

***
**Exercises**

1. Create a vector of random numbers using the code below:

```{r}
	random_numbers <- c(81, 90, 65, 43, 71, 29)
```

2. Use the pipe (`%>%`) to perform two steps in a single line:
	
* Take the mean of `random_numbers` using the `mean()` function.
* Round the output to three digits using the `round()` function.

***

### Tibbles


A core component of [tidyverse](http://tidyverse.org/) is the [tibble](http://tibble.tidyverse.org/). **Tibbles are data frames, but have several properties that make it superior. For example, printing the tibble to screen displays the data types in each of the columns and the dimensions. Another very handy feature of tibbles is that by default they will only print out the first 10 rows and as many columns as fit in your window. This is important when you are working with large datasets -- as we are.


Tibbles can be created directly using the `tibble()` function or data frames can be converted into tibbles using `as_tibble(name_of_df)`. It is also easy to convert `tibbles` into dataframes with the `as.data.frame()` function.


>**NOTE:** The function `as_tibble()` will ignore row names, so if a column representing the row names is needed, then the function `rownames_to_column(name_of_df)` should be run prior to turning the `data.frame` into a tibble. `rownames_to_column()` takes the rownames and adds it as a column in the data frame.

## Experimental data

We're going to explore the Tidyverse suite of tools to wrangle our data to prepare it for visualization. You should have `gprofiler_results_Mov10oe.tsv` in your R project's `data` folder earlier. 

**The dataset:**

* Represents the **functional analysis results**, including the biological processes, functions, pathways, or conditions that are over-represented in a given list of genes.

* Our gene list was generated by **differential gene expression analysis** and the genes represent differences between **control mice** and **mice over-expressing a gene involved in RNA splicing**. 

The functional analysis that we will focus on involves **gene ontology (GO) terms**, which:

* describe the roles of genes and gene products
* organized into three controlled vocabularies/ontologies (domains):
	* biological processes (BP)
	* cellular components (CC)
	* molecular functions (MF)

![Alt text](img/mov10_FA.png){ width=1200px }

## Analysis goal and workflow

**Goal:** *Visually compare the most significant biological processes (BP) based on the number of associated differentially expressed genes (gene ratios) and significance values by creating the following plot:*

![dotplot6](img/dotplot6.png)

To wrangle our data in preparation for the plotting, we are going to use the Tidyverse suite of tools to wrangle and visualize our data through several steps:

1. Read in the functional analysis results

2. Extract only the GO biological processes (BP) of interest

3. Select only the columns needed for visualization

4. Order by significance (p-adjusted values)

5. Rename columns to be more intuitive

6. Create additional metrics for plotting (e.g. gene ratios)

7. Plot results

**Tidyverse tools**

We are going to investigate more deeply the packages loaded with tidyverse, specifically the reading (`readr`), wrangling (`dplyr`), and plotting (`ggplot2`) tools.

### 1. Read in the functional analysis results

While the base R packages have perfectly fine methods for reading in data, the `readr` and `readxl` Tidyverse packages offer additional methods for reading in data. Let's read in our tab-delimited functional analysis results using `read_delim()`:

```{r}
# Read in the functional analysis results
functional_GO_results <- read_delim(file = "data/gprofiler_results_Mov10oe.tsv", delim = "\t" )

# Take a look at the results
functional_GO_results
```

```{r, results='asis', echo=FALSE}

collapsible_code(
  # HTML output: Collapsible section with raw HTML
  '<details>\n<summary><i>Click here to see how to do this in base R</i></summary>\n<pre><code># Read in the functional analysis results\nfunctional_GO_results <- read.delim(file = "data/gprofiler_results_Mov10oe.tsv", sep = "\\t")</code>\n<code># Take a look at the results\nfunctional_GO_results</code></pre>\n</details>',

  # Non-HTML output: Regular Markdown code block
  "### How to do this in base R\n\n```r\n# Read in the functional analysis results\nfunctional_GO_results <- read.delim(file = \"data/gprofiler_results_Mov10oe.tsv\", sep = \"\\t\")\n# Take a look at the results\nfunctional_GO_results\n```\n"
)
```

Notice that the results were automatically read in as a tibble and the output gives the number of rows, columns and the data type for each of the columns.


### 2. Extract only the GO biological processes (BP) of interest

Now that we have our data, we will need to wrangle it into a format ready for plotting.

To extract the biological processes of interest, we only want those rows where the `domain` is equal to `BP`, which we can do using the `filter()` function from the `dplyr` package. 

To filter rows of a data frame/tibble based on values in different columns, we give a logical expression as input to the `filter()` function to return those rows for which the expression is TRUE.

Now let's return only those rows that have a `domain` of `BP`:

```{r}
# Return only GO biological processes
bp_oe <- functional_GO_results %>%
  filter(domain == "BP")
  
head(bp_oe[,-7])

```



```{r, results='asis', echo=FALSE}

collapsible_code('<details>\n<summary><i>Click here to see how to do this in base R</i></summary>\n<pre><code># Return only GO biological processes\nidx <- functional_GO_results$domain == "BP"\nbp_oe2 <- functional_GO_results[idx,]</code>\n<code># View(bp_oe</code></pre>\n</details>',
"### How to do this in base R\n\n````{r}\n# Return only GO biological processes\nidx <- functional_GO_results$domain == \"BP\"\nbp_oe2 <- functional_GO_results[idx,]\n# View(bp_oe)\n````\n")

```

**Now we have returned only those rows with a `domain` of `BP`. **How have the dimensions of our results changed?**

***
**Exercise:**

We would like to perform an additional round of filtering to only keep the most specific GO terms. 

1. For `bp_oe`, use the `filter()` function to only keep those rows where the `relative.depth` is greater than 4.
2. Save output to overwrite our `bp_oe` variable.

***


### 3. Select only the columns needed for visualization

For visualization purposes, we are only interested in the columns related to the GO terms, the significance of the terms, and information about the number of genes associated with the terms. 

To extract these columns from a data frame/tibble we can use the `select()` function. In contrast to base R, we do not need to put the column names in quotes for selection.

```{r}
# Selecting columns to keep
bp_oe <- bp_oe %>%
  select(term.id, term.name, p.value, query.size, term.size, overlap.size, intersection)

head(bp_oe)
```

```{r, results='asis', echo=FALSE}

collapsible_code(
  # HTML content (collapsible section)
  '<details>\n<summary><i>Click here to see how to do this in base R</i></summary>\n<pre><code># Selecting columns to keep\nbp_oe <- bp_oe[, c("term.id", "term.name", "p.value", "query.size", "term.size", "overlap.size", "intersection")]\n# View(bp_oe)</code></pre>\n</details>',
  
  # Non-HTML content (R code block for EPUB/PDF)
  "### How to do this in base R\n\n```r\n# Selecting columns to keep\nbp_oe <- bp_oe[, c(\"term.id\", \"term.name\", \"p.value\", \"query.size\", \"term.size\", \"overlap.size\", \"intersection\")]\n# View(bp_oe)\n```\n"
)
```

	
![Alt text](img/bp_oe_selection.png){ width=1200px }



### 4. Order GO processes by significance (adjusted p-values)

Now that we have only the rows and columns of interest, let's arrange these by significance, which is denoted by the adjusted p-value.

Let's sort the rows by adjusted p-value with the `arrange()` function from the `dplyr` package.

```{r}
# Order by adjusted p-value ascending
bp_oe <- bp_oe %>% arrange(p.value)
```
### 5. Rename columns to be more intuitive

While not necessary for our visualization, renaming columns more intuitively can help with our understanding of the data using the `rename()` function from the `dplyr` package. The syntax is `new_name` = `old_name`.


Let's rename the `term.id` and `term.name` columns.

```{r}
# Provide better names for columns
bp_oe <- bp_oe %>% dplyr::rename(GO_id = term.id, 
                GO_term = term.name)
```


```{r, results='asis', echo=FALSE}

collapsible_code('<details>\n<summary><i>Click here to see how to do this in base R</i></summary>\n<pre><code># Provide better names for columns\ncolnames(bp_oe)[colnames(bp_oe) == "term.id"] <- "GO_id"\ncolnames(bp_oe)[colnames(bp_oe) == "term.name"] <- "GO_term"</code></pre>\n</details>',
"### How to do this in base R\n\n```r\n# Provide better names for columns\ncolnames(bp_oe)[colnames(bp_oe) == \"term.id\"] <- \"GO_id\"\ncolnames(bp_oe)[colnames(bp_oe) == \"term.name\"] <- \"GO_term\"\n```\n")

```
	
> **NOTE:**  In the case of two packages with identical function names, you can use `::` with the package name before and the function name after (e.g `stats::filter()`) to ensure that the correct function is implemented. The `::` can also be used to bring in a function from a library without loading it first.
> 
> In the example above, we wanted to use the `rename()` function specifically from the `dplyr` package, and not any of the other packages (or base R) which may have the `rename()` function.

***
**Exercise**

Rename the `intersection` column to `genes` to reflect the fact that these are the DE genes associated with the GO process.

***

### 6. Create additional metrics for plotting (e.g. gene ratios)

Finally, before we plot our data, we need to create a couple of additional metrics. The `mutate()` function enables you to create a new column from an existing column. 

Let's generate gene ratios to reflect the number of DE genes associated with each GO process relative to the total number of DE genes.

```{r}
# Create gene ratio column based on other columns in dataset
bp_oe <- bp_oe %>%
  mutate(gene_ratio = overlap.size / query.size)
```

```{r, results='asis', echo=FALSE}

collapsible_code('<details>\n<summary><i>Click here to see how to do this in base R</i></summary>\n<pre><code># Create gene ratio column based on other columns in dataset\nbp_oe <- cbind(bp_oe, gene_ratio = bp_oe$overlap.size / bp_oe$query.size)</code></pre>\n</details>',
"### How to do this in base R\n\n```r\n# Create gene ratio column based on other columns in dataset\nbp_oe <- cbind(bp_oe, gene_ratio = bp_oe$overlap.size / bp_oe$query.size)\n```\n")

```
	
***
**Exercise**

Create a column in `bp_oe` called `term_percent` to determine the percent of DE genes associated with the GO term relative to the total number of genes associated with the GO term (`overlap.size` / `term.size`)

***

Our final data for plotting should look like the table below:

![Alt text](img/bp_oe_ready_to_plot.png){ width=1200px }

## Next steps

Now that we have our results ready for plotting, we can use the [ggplot2](https://ggplot2.tidyverse.org) package to plot our results. 

**Homework**

Use ggplot2 to plot bp_oe to recapitulate the GO enrichment figure shown below using the top 30 categories:

![dotplot6](img/dotplot6.png)


### Additional resources

* [R for Data Science](http://r4ds.had.co.nz)
* [teach the tidyverse](http://varianceexplained.org/r/teach-tidyverse/)
* [tidy style guide](http://style.tidyverse.org/)

***
